### PR TITLE
ENH - Better error reporting when initializing classes

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
This provides details about errors that may occur during the initialization of classes.

### Preview of changes
Running the example in ``reproduce_benchmark`` folder using ``main`` then current branch 

<details>

<summary>before changes</summary>

```bash
...
    all_solvers = _list_all_parametrized_names(*self.get_solvers())
  File "/home/badr-moufad/Desktop/benchmarks-repo/benchopt/benchopt/utils/parametrized_name_mixin.py", line 127, in _list_all_parametrized_names
    cls._get_parametrized_name(**dataset_parameters)
  File "/home/badr-moufad/Desktop/benchmarks-repo/benchopt/benchopt/utils/parametrized_name_mixin.py", line 73, in _get_parametrized_name
    return str(cls.get_instance(**parameters))
  File "/home/badr-moufad/Desktop/benchmarks-repo/benchopt/benchopt/utils/parametrized_name_mixin.py", line 39, in get_instance
    obj = cls(**parameters)
TypeError: __init__() got an unexpected keyword argument 'param_2'
```
</details>

<details>

<summary>after changes</summary>

```bash
...
    all_solvers = _list_all_parametrized_names(*self.get_solvers())
  File "/home/badr-moufad/Desktop/benchmarks-repo/benchopt/benchopt/utils/parametrized_name_mixin.py", line 126, in _list_all_parametrized_names
    cls._get_parametrized_name(**dataset_parameters)
  File "/home/badr-moufad/Desktop/benchmarks-repo/benchopt/benchopt/utils/parametrized_name_mixin.py", line 72, in _get_parametrized_name
    return str(cls.get_instance(**parameters))
  File "/home/badr-moufad/Desktop/benchmarks-repo/benchopt/benchopt/utils/parametrized_name_mixin.py", line 39, in get_instance
    obj = cls(**parameters)
TypeError: Error when initializing `Solver: gradient-descent`. __init__() got an unexpected keyword argument 'param_2'
```

</details>

-------
P.S. The ``reproduce_bench`` folder gives a succinct example to reproduce these behaviors and shall be removed afterward.

### CODE REVIEW

File: benchopt/utils/parametrized_name_mixin.py [hunk 1]
```diff
@@ -38,9 +38,19 @@ def get_instance(cls, **parameters):
...
+        except Exception as exception:
+            # get type (Dataset, Objective, or Solver) of class and its name
+            cls_type = cls.__base__.__name__
+            cls_type = cls_type[4:]  # remove Base
+            cls_name = cls.name
+
+            # Extend exception error message
+            # TODO: use `add_note` when requiring python>=3.11
+            exception.args = (
+                f"Error when initializing `{cls_type}: {cls_name}`. "
```
Comment (tomMoral) [inline benchopt/utils/parametrized_name_mixin.py:50]:
```suggestion
                f'Error when initializing {cls_type} "{cls_name}". '
```

**Review by tomMoral — COMMENTED** [review]
Could you add a test where you pass the arguments through the command lines with a non existing argument? This would allow testing this functionality without adding a new bench.

File: benchopt/utils/parametrized_name_mixin.py [hunk 2]
```diff
@@ -38,9 +38,19 @@ def get_instance(cls, **parameters):
...
+                f"{'. '.join(exception.args)}",)
```
[the remark this answers is not in the document]
Comment (Badr-MOUFAD) [reply benchopt/utils/parametrized_name_mixin.py]: > Maybe ti would be sufficient to just print the message before raising the error? 
It is possible to proceed this way, but I think the limitation is that the prints are put at the top of the error trace hence they are difficult to identify

> Or would this mess up with the is_installed functionalities
I will check the unittest

**Comment by Badr-MOUFAD** [comment]
> Could you add a test where you pass the arguments through the command lines with a non existing argument? This would allow testing this functionality without adding a new bench.

I unit-tested the code changes. Yet, I don't think that it should go in the ``test_cli.py`` as we enhance error reporting during the implementation of an `Objective`, `Dataset` or `Solver`

File: benchopt/tests/test_cli.py [hunk 3]
```diff
@@ -348,6 +348,46 @@ class Solver(BaseSolver):
...
+            "    def __init__(self, param=1., deprecated_return=False):\n"
+            "        self.param = param\n"
+            "        self.deprecated_return = deprecated_return\n"
+            "    def get_data(self):\n"
+            "        X = [[1, 0], [0, 1]]\n"
+            "        y = [1, 1]\n"
+            "        return dict(X=X, y=y)\n"
+        )
+        dataset_tmp = tempfile.NamedTemporaryFile(
+            mode="w+", suffix='.py', dir=DUMMY_BENCHMARK_PATH / "datasets")
```
Comment (tomMoral) [inline benchopt/tests/test_cli.py:367]:
```suggestion
            mode="w+", suffix='.py', dir=DUMMY_BENCHMARK_PATH / "datasets"
        )
```

File: benchopt/tests/test_cli.py [hunk 4]
```diff
@@ -348,6 +348,46 @@ class Solver(BaseSolver):
...
+            "class Dataset(BaseDataset):\n"
+            "    name = 'buggy-dataset'\n"
+            "    parameters = {'wrong_param_name': [100]}\n"
+            "    def __init__(self, param=1., deprecated_return=False):\n"
+            "        self.param = param\n"
+            "        self.deprecated_return = deprecated_return\n"
+            "    def get_data(self):\n"
+            "        X = [[1, 0], [0, 1]]\n"
+            "        y = [1, 1]\n"
+            "        return dict(X=X, y=y)\n"
```
Comment (tomMoral) [inline benchopt/tests/test_cli.py:364]:
```suggestion
```

File: benchopt/tests/test_cli.py [hunk 5]
```diff
@@ -348,6 +348,46 @@ class Solver(BaseSolver):
...
+    def test_handle_class_init_error(self):
+        # dataset with a wrong param name
+        dataset_src = (
+            "from benchopt import BaseDataset\n"
+            "class Dataset(BaseDataset):\n"
+            "    name = 'buggy-dataset'\n"
+            "    parameters = {'wrong_param_name': [100]}\n"
+            "    def __init__(self, param=1., deprecated_return=False):\n"
+            "        self.param = param\n"
+            "        self.deprecated_return = deprecated_return\n"
```
Comment (tomMoral) [inline benchopt/tests/test_cli.py:360]:
```suggestion
```

File: benchopt/tests/test_cli.py [hunk 6]
```diff
@@ -348,6 +348,46 @@ class Solver(BaseSolver):
...
+        dataset_tmp.flush()
+
+        config = f"""
+            objective-filter:
+            - {SELECT_ONE_OBJECTIVE}
+            dataset:
+            - buggy-dataset
+            max-runs: 1
+            solver:
+            - python-pgd[step_size=[2, 3]]
```
Comment (tomMoral) [inline benchopt/tests/test_cli.py:378]:
```suggestion
            - python-pgd[step_size=2]
```

**Review by tomMoral — COMMENTED** [review]
That is pretty nice. I think the test in `test_cli.py` is okay as this informs about how well the CLI is reporting error to the users.

A few suggestion to simplify the unittest.
Also maybe consider putting  the tempfiles as context to ensure that there is no border effects on other tests?

File: benchopt/tests/test_cli.py (hunk 4 above)
Comment (Badr-MOUFAD) [reply benchopt/tests/test_cli.py]: It's mandatory to retain the `get_data` method as it's an abstract method of `BaseDataset`.
I kept ``return dict()``.

**Comment by Badr-MOUFAD** [comment]
> Also maybe consider putting the tempfiles as context to ensure that there is no border effects on other tests?

I tried that but I get a syntax error in the `as` keyword which I can't reproduce locally. Any workaround?

File: benchopt/tests/test_cli.py (hunk 4 above)
Comment (tomMoral) [reply benchopt/tests/test_cli.py]: perfect. thanks

**Comment by tomMoral** [comment]
> I tried that but I get a syntax error in the `as` keyword which I can't reproduce locally. Any workaround?

It is an issue due to this: https://stackoverflow.com/a/1073814/2642845
Prior to python3.10, you cannot use parenthesis in the `with` statement. You need to use backslash.

**Review by mathurinm — APPROVED** [review]

File: benchopt/tests/test_cli.py [hunk 7]
```diff
@@ -348,6 +348,47 @@ class Solver(BaseSolver):
...
+            tmp_dataset.write(dataset_src)
+            tmp_dataset.flush()
+
+            tmp_config.write(config)
+            tmp_config.flush()
+
+            run_cmd = [str(DUMMY_BENCHMARK_PATH), '--config', tmp_config.name,
+                       '--no-plot']
+
+            error_match = r"""Dataset: "buggy-dataset".*'wrong_param_name'"""
```
Comment (tomMoral) [inline benchopt/tests/test_cli.py:388]:
```suggestion
            error_match = """Dataset: "buggy-dataset".*'wrong_param_name'"""
```

**Review by tomMoral — APPROVED** [review]
LGTM, thx @Badr-MOUFAD

**Pull request merged**